src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/NewAFCalculator.java

```diff
@@ -0,0 +1,165 @@
+     * @return result (for programming convenience)
+     */
+    @Override
+    public AFCalculationResult getLog10PNonRef(final VariantContext vc, final int defaultPloidy, final int maximumAlternativeAlleles, final double[] refSnpIndelPseudocounts) {
+        Utils.nonNull(vc, "VariantContext cannot be null");
+        final int numAlleles = vc.getNAlleles();
+        final List<Allele> alleles = vc.getAlleles();
+        Utils.validateArg( numAlleles > 1, () -> "VariantContext has only a single reference allele, but getLog10PNonRef requires at least one at all " + vc);
+
+        final double[] priorPseudocounts = alleles.stream()
+                .mapToDouble(a -> a.isReference() ? refPseudocount : (a.length() > 1 ? snpPseudocount : indelPseudocount)).toArray();
+
+        final Dirichlet prior = new Dirichlet(priorPseudocounts);
+        double[] alleleCounts = new double[numAlleles];
+        final double flatLog10AlleleFrequency = -MathUtils.log10(numAlleles); // log10(1/numAlleles)
+        double[] log10AlleleFrequencies = new IndexRange(0, numAlleles).mapToDouble(n -> flatLog10AlleleFrequency);
+        double alleleCountsMaximumDifference = Double.POSITIVE_INFINITY;
+
+        while (alleleCountsMaximumDifference > THRESHOLD_FOR_ALLELE_COUNT_CONVERGENCE) {
+            final double[] newAlleleCounts = effectiveAlleleCounts(vc, log10AlleleFrequencies);
+            alleleCountsMaximumDifference = Arrays.stream(MathArrays.ebeSubtract(alleleCounts, newAlleleCounts)).map(Math::abs).max().getAsDouble();
+            alleleCounts = newAlleleCounts;
+
+            // first iteration uses flat prior in order to avoid local minimum where the prior + no pseudocounts gives such a low
+            // effective allele frequency that it overwhelms the genotype likelihood of a real variant
+            // basically, we want a chance to get non-zero pseudocounts before using a prior that's biased against a variant
+            log10AlleleFrequencies = new Dirichlet(prior, alleleCounts).log10MeanWeights();
+        }
+
+        double[] log10POfZeroCountsByAllele = new double[numAlleles];
+        double log10PNoVariant = 0;
+
+        for (final Genotype g : vc.getGenotypes()) {
+            if (!g.hasLikelihoods()) {
+                continue;
+            }
+            final int ploidy = g.getPloidy() == 0 ? defaultPloidy : g.getPloidy();
+            final GenotypeLikelihoodCalculator glCalc = GL_CALCS.getInstance(ploidy, numAlleles);
+            final double[] genotypePosteriors = normalizedGenotypePosteriors(g, glCalc, log10AlleleFrequencies);
+
+            //the total probability
+            log10PNoVariant += Math.log10(genotypePosteriors[HOM_REF_GENOTYPE_INDEX]);
+
+            // per allele non-log space probabilities of zero counts for this sample
+
+
+            // for each allele calculate the total probability of genotypes containing at least one copy of the allele
+            double[] pOfNonZeroAltAlleles = new double[numAlleles];
```

Comment:
shouldn't pOfNonZeroAltAlleles be final? actually I think is surprising that you can access it from within the lambda below if it is not final, but it must be compiling.